### PR TITLE
Java Client 9.0.0 for JDK21

### PR DESCRIPTION
**Pull request by BrianNichols**
Java Client 9.0.0 for JDK21

### CODE REVIEW

File: client/src/com/aerospike/client/command/Command.java
Comment (rashidmayes-aero) [inline client/src/com/aerospike/client/command/Command.java:1]: I can see some compression in the write command when using the latest Java client. I have walked through the code and the compressed buffer is placed on the wire. However, I am not seeing the corresponding record_proto_compression_ratio change on the server. The ratio is always 1.

image


 


@tfaulkes @BrianNichols
Is this the correct metric to monitor to confirm that client compression is working?

Comment (BrianNichols) [reply client/src/com/aerospike/client/command/Command.java:1]: I'm not familiar with server compression statistic details. @kportertx Do you know why the compression ratio is unchanged on the server?

Comment (kportertx) [reply client/src/com/aerospike/client/command/Command.java:1]: 1 would either mean that there isn't any wire compression or that the compression isn't helping.

Comment (rashidmayes-aero) [reply client/src/com/aerospike/client/command/Command.java:1]: ok, I will do more investigation and test against other clusters.

I have this record:

`				"bins": {
						"a": "00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000",
						"partner":  "1212121212" 
				}`


On the client, it reduces from 1053 (dataOffset) to 125 (csize).  
client conn write:

image

image

Yet, the server stats show

image

The compression is helping, at least appears so through the client write. The server storage compression ratio is significant also.

Comment (kportertx) [reply client/src/com/aerospike/client/command/Command.java:1]: You could measure the number of bytes your nic sent/received using the ip shell command. Run a workload with/without compression enabled on the client and verify you send or receive less bytes with compression.

You could also use Wireshark.

**Comment by rashidmayes-aero** [comment]
Did some testing using Wireshark, etc. I looks like asd is receiving the compressed data (replica write seems uncompressed??). Here are the results:

### asd Host VM

**Uncompressed:**

Screenshot 2026-01-09 at 12 14 26

**Compressed:**

Screenshot 2026-01-09 at 12 13 17


### Mac


**Uncompressed:**

Screenshot 2026-01-09 at 11 51 41

**Compressed**


Screenshot 2026-01-09 at 11 53 08


**Uncompressed to Compressed Dropoff**

Screenshot 2026-01-09 at 11 52 24


### Wireshark

**Ucompressed**

Screenshot 2026-01-09 at 12 02 02


**Compressed**

Screenshot 2026-01-09 at 12 03 32


### Server stats

Screenshot 2026-01-09 at 00 46 39

File: client/src/com/aerospike/client/command/Command.java
Comment (rashidmayes-aero) [reply client/src/com/aerospike/client/command/Command.java:1]: Did some testing using Wireshark, etc. I looks like asd is receiving the compressed data (replica write seems uncompressed??). Here are the results:

### asd Host VM

**Uncompressed:**

Screenshot 2026-01-09 at 12 14 26

**Compressed:**

Screenshot 2026-01-09 at 12 13 17


### Mac


**Uncompressed:**

Screenshot 2026-01-09 at 11 51 41

**Compressed**


Screenshot 2026-01-09 at 11 53 08


**Uncompressed to Compressed Dropoff**

Screenshot 2026-01-09 at 11 52 24


### Wireshark

**Ucompressed**

Screenshot 2026-01-09 at 12 02 02


**Compressed**

Screenshot 2026-01-09 at 12 03 32


### Server stats

Screenshot 2026-01-09 at 00 46 39

Comment (rashidmayes-aero) [reply client/src/com/aerospike/client/command/Command.java:1]: I spoke with Meher and believe I have an answer. The metric only tracks the response to the client. I was testing only writes. I will do more testing with reads, gets and queries, to see if it changes.

Comment (rashidmayes-aero) [reply client/src/com/aerospike/client/command/Command.java:1]: Completed testing. Confirmed that the reads (data sent from server) are tracked. 

image

There does not appear to be appear to be an equivalent for incoming data.  Thanks for the help on this.

**Pull request merged**